pystiche_papers/sanakoyeu_et_al_2018/_modules.py

```diff
@@ -184,3 +192,202 @@ def residual_block(channels: int) -> ResidualBlock:
+        )
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        if self.impl_params:
+            return cast(torch.Tensor, self.forwardBlock(input))
+        else:
+            return cast(torch.Tensor, nn.utils.weight_norm(self.forwardBlock(input)))
+
+
+def discriminator_encoder_modules(
+    in_channels: int = 3, inplace: bool = True,
+) -> Dict[str, nn.Sequential]:
+    # FIXME:  if/when the Python interpreter will learn to accept the correct signature
+    # https://stackoverflow.com/questions/41207128/how-do-i-specify-ordereddict-k-v-types-for-mypy-type-annotation
+
+    modules = OrderedDict(
+        {
+            "scale_0": ConvBlock(
+                in_channels,
+                128,
+                kernel_size=5,
+                stride=2,
+                padding="same",
+                act="lrelu",
+                inplace=inplace,
+            ),
+            "scale_1": ConvBlock(
+                128,
+                128,
+                kernel_size=5,
+                stride=2,
+                padding="same",
+                act="lrelu",
+                inplace=inplace,
+            ),
+            "scale_2": ConvBlock(
+                128,
+                256,
+                kernel_size=5,
+                stride=2,
+                padding="same",
+                act="lrelu",
+                inplace=inplace,
+            ),
+            "scale_3": ConvBlock(
+                256,
+                512,
+                kernel_size=5,
+                stride=2,
+                padding="same",
+                act="lrelu",
+                inplace=inplace,
+            ),
+            "scale_4": ConvBlock(
+                in_channels=512,
+                out_channels=512,
+                kernel_size=5,
+                stride=2,
+                padding="same",
+                act="lrelu",
+                inplace=inplace,
+            ),
+            "scale_5": ConvBlock(
+                512,
+                1024,
+                kernel_size=5,
+                stride=2,
+                padding="same",
+                act="lrelu",
+                inplace=inplace,
+            ),
+            "scale_6": ConvBlock(
+                1024,
+                1024,
+                kernel_size=5,
+                stride=2,
+                padding="same",
+                act="lrelu",
+                inplace=inplace,
+            ),
+        }
+    )
+    return cast(Dict[str, nn.Sequential], modules)
+
+
+class DiscriminatorEncoder(MultiLayerEncoder):
+    r"""Encoder part of the Discriminator from :cite:`SKL+2018`.
+
+    Args:
+        in_channels: Number of channels in the input. Defaults to ``3``.
+    """
+
+    def __init__(self, in_channels: int = 3) -> None:
+        super().__init__(
+            self._collect_modules(
+                discriminator_encoder_modules(in_channels=in_channels)
+            )
+        )
+
+    def _collect_modules(
```

Comment:
I'm confused: why are you first naming the blocks `scale_x` just to rename them here?

Reply:
As I understood the `MultiLayerEncoder`, it does not work with a `nn.Sequential`. But since the `ConvBlock`s are a `nn.Sequence` of convolution, normalization and activation, I first named the `ConvBlock`s as in the paper with `scale_x` and then I brought it into `DiscriminatorEncoder` into the structure of the `MultiLayerEncoder`.

Reply:
I've never tried this, but in theory there is nothing that prevents us from using sequentials within an `MultiLayerEncoder`. If we do that each sequential is treated as a single layer, i.e. we can't access the inner encodings. But AFAIK this is not needed here, is it?

Reply:
Only the encodings after activation are needed here. So you are right in your assumption that the inner encodings are not needed.

Reply:
In that case, just use the sequentials as layers.